### PR TITLE
Add maxDegreeOfParallelism to Async.Parallel

### PR DESCRIPTION
**Pull request by Frassle**
Implementation idea for https://github.com/fsharp/fslang-suggestions/issues/706

### CODE REVIEW

**Comment by dsyme** [comment]
This would break binary compat of FSharp.Core - it would need to be added as a separate overload.

We should also have a small RFC for this I suppose.

**Comment by dsyme** [comment]
(I like the change)

**Comment by Frassle** [comment]
Yes the whitespace changes were committed by mistake. I will revert those. Am also going to add a few more tests for parameter validation and checking that exceptions and cancellations behave correctly.

**Comment by cartermp** [comment]
@Frassle Thanks!

**Comment by Frassle** [comment]
Updated. CI is currently failing because of https://github.com/Microsoft/visualfsharp/pull/6360 so the tests don't see the new Parallel overload. 
I wasn't sure about the .xlf files, I assumed the generated updates should be committed?

File: src/fsharp/FSharp.Core/async.fs [hunk 1]
```diff
@@ -1281,35 +1281,33 @@ namespace Microsoft.FSharp.Control
...
-                                                    (fun edi -> recordFailure (Choice1Of2 edi))
-                                                    (fun cexn -> recordFailure (Choice2Of2 cexn))
-                                            tasks.[j].Invoke ctxt
-                                        )
-                                        |> unfake
-                                fake()
-                            )
+                        let worker = MakeAsync (fun _ ->
+                            while i < tasks.Length do
+                                let j = Interlocked.Increment &i - 1
```
Comment (saul) [inline src/fsharp/FSharp.Core/async.fs]: Alternatively, could `let mutable i = -1` (and not have to `- 1` here)

**Comment by NatElkins** [comment]
Just going to throw this here,...Jet has a bunch of Async extensions related to this feature.  Might be worth taking a look and comparing the implementation: https://github.com/jet/kafunk/blob/master/src/kafunk/Utility/Async.fs

**Comment by Frassle** [comment]
This implementation is quite different, it does limiting the same way TPL does limiting by starting _maxDegreeOfParallelism_ workers, rather than looping to enter a semaphore. Should be quite a lot more efficient, spinning up less threads.

I did think about _startOnCallingThread_, especially for _Sequential_ maybe it should be written to not queue at the start, but even for _Parallel_ could set it up so one worker just starts on the current thread rather than all being queued. I wasn't sure what the guidelines we're for when asyncs should try to reuse the current thread, or if it should be a parameter.

**Comment by KevinRansom** [comment]
Needs to go in F# 5

**Comment by Frassle** [comment]
Do I need to point this at another branch for that?

**Comment by KevinRansom** [comment]
We'll take care of it :-)

**Comment by cartermp** [comment]
@KevinRansom This was still failing surface area tests.

**Comment by KevinRansom** [comment]
I know, I have a fix.

**Comment by KevinRansom** [comment]
Here: //github.com/dotnet/fsharp/commit/14646945a98288b221a109fbee49bae34ef57fc3

**Pull request merged**